crates/typst-syntax/src/ast.rs

```diff
@@ -1,6 +1,83 @@
+Typst's Abstract Syntax Tree (AST) is as a lazy, typed view over the untyped
+Concrete Syntax Tree (CST) and is rooted in the [`Markup`] node.
+
+## The AST is a View
+
+Most AST nodes are wrapper structs around [`SyntaxNode`] pointers. This summary
+will use a running example of the [`Raw`] node type, which is declared (after
+macro expansion) as: `struct Raw<'a>(&'a SyntaxNode);`.
+
+[`SyntaxNode`]s are generated by the parser and constitute the Concrete Syntax
+Tree (CST). The CST is _concrete_ because it has the property that an in-order
+tree traversal will recreate the text of the source file exactly.
+
+[`SyntaxNode`]s in the CST contain their [`SyntaxKind`], but don't themselves
+provide access to the semantic meaning of their contents. That semantic meaning
+is available throught the Abstract Syntax Tree by iterating over CST nodes and
+inspecting their contents. The format is prepared ahead-of-time by the parser so
+that this module can unpack the abstract meaning from the CST's structure.
+
+Raw nodes are parsed by recognizing paired backtick delimiters, which you will
+find as CST nodes with the [`RawDelim`] kind. However, the AST doesn't include
+these delimiters because it _abstracts_ over the backticks. Instead, the parent
+raw node will only use its child [`RawDelim`] CST nodes to determine whether the
+element is a block or inline.
+
+## The AST is Typed
+
+The AST gives access to methods on nodes of the tree. The [`Raw`] node provides
+3 methods that specify its abstract representation: [`Raw::lines()`] returns the
+raw text as an iterator of lines, [`Raw::lang()`] provides the optionally
+present [`RawLang`] language tag, and [`Raw::block()`] gives a bool for whether
+the raw element is a block or inline.
+
+These methods are implemented by assuming the structure of [`SyntaxNode`]s will
+match what is generated by the parser. The [`Raw`] struct really is just a
+wrapper over a [`SyntaxNode`] reference. But to construct the wrapper and access
+the methods, we must check that the [`SyntaxKind`] of the underlying node
+matches.
+
+And this is a safe interface because the Rust compiler only gives access to the
+methods after a [`Raw`] struct is created via [`AstNode::from_untyped`]. The
+`node!` macro implements `from_untyped` by checking the node's kind before
+constructing it, returning `Some()` only if the kind matches the desired struct.
+
+## The AST is Lazy
+
+Being lazy means that the untyped CST nodes are converted to typed AST Nodes
+only as the tree is traversed. If we parse a file and a raw block is contained
+in a branch of an if-statement that we don't take, then we won't pay the cost of
+creating an iterator over the lines or checking whether it was a block or
+inline (although it will still be parsed into nodes). This is also a factor of
+the current "tree-interpreter" evaluation model. A bytecode interpreter might
+instead eagerly convert the AST into bytecode, but it will still traverse using
+this lazy interface.
```

Comment:
Debatable, but would be nice to give more insight into this difference:

```suggestion
Being lazy means that the untyped CST nodes are converted to typed AST Nodes
only as the tree is traversed. If we parse a file and a raw block is contained
in a branch of an if-statement that we don't take, then we won't pay the cost of
creating an iterator over the lines or checking whether it was a block or
inline (although it will still be parsed into nodes).

This is also a factor of
the current "tree-interpreter" evaluation model. A bytecode interpreter might
instead eagerly convert the AST into bytecode, but it will still traverse using
this lazy interface. The advantage, in that case, is that it would only traverse
through the same node once. The current interpreter has to traverse it multiple
times to execute the same code. However, it is also more flexible and easier to
develop upon.
```

Reply:
I rewrote this to the following:
```markdown
## The AST is Lazy

Being lazy means that the untyped CST nodes are converted to typed AST nodes
only as the tree is traversed. If we parse a file and a raw block is contained
in a branch of an if-statement that we don't take, then we won't pay the cost of
creating an iterator over the lines or checking whether it was a block or
inline (although it will still be parsed into nodes).

This is also a factor of the current "tree-interpreter" evaluation model. A
bytecode interpreter might instead eagerly convert the AST into bytecode, but it
would still traverse using this lazy interface. While the tree-interpreter
evaluation is straightforward and easy to add new features onto, it has to
re-traverse the AST every time a function is evaluated. A bytecode interpreter
using the lazy interface would only need to traverse each node once, improving
throughput at the cost of initial latency and development flexibility.
```